File: arc166/src/bin/b.rs
use num_integer::Integer;
use proconio::input;

fn main() {
    input! {
        n: usize,
        dd: [usize; 3],
        aa: [usize; n],
    }

    let mut dp = vec![10_usize.pow(18); 8];
    dp[0] = 0;
    for &a in &aa {
        let mut next_dp = dp.clone();

        for bit in 0..7 {
            let rem = 7 ^ bit;
            let mut add = rem;
            while add != 0 {
                let d = (0..3)
                    .filter(|&i| (add >> i) & 1 == 1)
                    .fold(1, |acc, i| acc.lcm(&dd[i]));
                next_dp[bit | add] = next_dp[bit | add].min(dp[bit] + (d - a % d) % d);

                add = (add - 1) & rem;
            }
        }

        dp = next_dp;
    }

    println!("{}", dp[7]);
}

File: arc166/b.py
import math
a=lambda:map(int,input().split())
_,*b=a()
c=[0]+[2**60]*7
for d in a():
 e=7
 while e:
  e-=1;f=7
  while f:g=math.lcm(*(b[i&3]for i in(1,2,4)if f&i));c[e|f]=min(c[e|f],c[e]+(g-d%g)%g);f-=1
print(c[7])